SW/Boiling Detection/boil_detect.py

```diff
@@ -0,0 +1,240 @@
+        else:
+            return False
+
+    def __burners(self,stove):
+        """
+        Returns a list of the Burners in the Stove object
+        :param stove: Stove object
+        :return: List of the burners
+        """
+        burners = []
+        burners.append(stove.upper_left)
+        burners.append(stove.upper_right)
+        burners.append(stove.lower_left)
+        burners.append(stove.lower_right)
+        return burners
+
+    def __initialize_masks(self,burners):
+        """
+        Refreshes mask if a pot is newly detected on the burners.
+        :param burners: Stove burners
+        """
+        masks = self.masks
+        for mask,burner,quad in zip(masks,burners,self.first_quads):
+            # mask never made before
+            if mask is None and quad is not None and burner.pot_detected:
+                self.masks.append(self.__create_mask(quad))
+            # if pot no longer detected resets the mask
+            elif burner.pot_detected is False:
+                idx = masks.index(mask)
+                self.masks[idx] = None
+
+    def update(self, frame, stove):
+        """
+        Takes an input frame and updates the boiling state of the burners.
+        :param frame: Input frame used to update boiling status
+        :param stove: Stove object holding current stove state
+        """
+        # updates frame_count
+        self.frame_count = self.frame_count + 1
+        burners = self.__burners(stove)
+
+        # creates new masks if a pot was added
+        self.initialize_masks(burners)
+
+        # if program has not executed before
+        if self.first_image is None:
+            self.first_image = frame
+            self.first_quads = self.__split_frame(self.first_image)
+
+        # if the number of frames that have been processed are desired amount, proceed
+        elif self.frame_count == self.count:
+            # if there is no second image recorded yet
+            if self.second_image is None:
+                self.second_image = frame
+                self.second_quads = self.__split_frame(self.second_image)
+            # second image does exist, must shift images
+            else:
+                self.first_image = self.second_image
+                self.second_image = frame
+
+            # splits the frames
+            self.first_quads = self.__split_frame(self.first_image)
+            self.second_quads = self.__split_frame(self.second_image)
+
+            # Detects boiling on active burners
+            i = 0
+            for burner in burners:
+                # if there is a pot on burner
+                if burner.pot_detected:
+                    mask = self.masks[i]
+                    score = self.___get_score(mask, self.first_quads[i], self.second_quads[i])
+                    self.median_SSIM[i].append(score)
+
+                    # median is calculated when enough SSIMs have been recorded
+                    if len(self.median_SSIM[i]) == self.SSIM_count:
```

Comment:
talk to mona -- she's gonna implement a StreamingAverage class that will help like this.

This means we can update more often while also taking averages over a window.

It's really just a queue. Every loop it pops an element (if the length is > then the window) and adds a new reading. Every loop, it also takes an average across the whole queue.

Reply:
Yes, it will be included in the main file so I'll let you know once I implement it and test it and then you can use it.

Reply:
Included this in the most recent commits